// @rallycommerce/swell/rollup.config.js
import summary from 'rollup-plugin-summary';
import { terser } from 'rollup-plugin-terser';
import replace from '@rollup/plugin-replace';
import resolve from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import typescript from '@rollup/plugin-typescript';
import del from 'rollup-plugin-delete';
import { environment } from './environment';
import babel from '@rollup/plugin-babel';

const env = environment(process.env.configuration);

export default {
  input: 'src/index.ts',
  output: [
    {
      name: 'RallySwellCheckoutButton',
      file: `build/index.umd.js`,
      format: 'umd'
    },
    {
      file: `build/index.cjs.js`,
      format: 'cjs',
    },
    {
      file: `build/index.esm.js`,
      format: 'es'
    }
  ],
  onwarn(warning) {
    if (warning.code !== 'THIS_IS_UNDEFINED') {
      console.error(`(!) ${warning.message}`);
    }
  },
  plugins: [
    del({ targets: 'build/*' }),
    replace({
      'Reflect.decorate': 'undefined',
      preventAssignment: true
    }),
    resolve(),
    commonjs(),
    typescript({
      tsconfig: './tsconfig.json',
      sourceMap: false,
      declarationDir: './'
    }),
    babel({
      babelHelpers: 'bundled'
    }),
    env.production && terser({
      format: {
        comments: false,
      },
    }),
    summary(),
  ],
  preserveEntrySignatures: 'strict',
};
